### PR TITLE
[RFC] Remove project-specific integer types: long_u. (4)

### PR DESCRIPTION
**Pull request by elmart**
Continues #1715.

Removes 60 long_u instances.
44 still remaining for future PR's.

### CODE REVIEW

File: src/nvim/hardcopy.c [hunk 1]
```diff
@@ -503,7 +504,7 @@ static void prt_header(prt_settings_T *psettings, int pagenum, linenr_T lnum)
   if (prt_use_number())
     width += PRINT_NUMBER_WIDTH;
 
-  tbuf = xmalloc(width + IOSIZE);
+  tbuf = xmalloc((size_t)width + IOSIZE);
```
Comment (fwalch) [inline src/nvim/hardcopy.c]: `width` is int, so should there be an assertion `>= 0` here?

Comment (elmart) [inline src/nvim/hardcopy.c]: Right :+1:

**Comment by elmart** [comment]
Travis build failed because of two warnings issued by gcc that are not issued by clang.
Will fix that and @fwalch suggestion and repush.

**Comment by elmart** [comment]
Fixed and repushed.

File: src/nvim/hardcopy.c [hunk 2]
```diff
@@ -455,8 +454,10 @@ static void prt_line_number(prt_settings_T *psettings, int page_line, linenr_T l
  */
 int prt_header_height(void)
 {
-  if (printer_opts[OPT_PRINT_HEADERHEIGHT].present)
-    return printer_opts[OPT_PRINT_HEADERHEIGHT].number;
+  if (printer_opts[OPT_PRINT_HEADERHEIGHT].present) {
+    assert(printer_opts[OPT_PRINT_HEADERHEIGHT].number <= INT_MAX);
```
Comment (fwalch) [inline src/nvim/hardcopy.c]: I'm pretty sure it's overkill, but if you want to be very strict, you would need to check for `printer_opts[OPT_PRINT_HEADERHEIGHT].number >= INT_MIN` as well, or change `option_table_T.number` to an unsigned type.

File: src/nvim/hardcopy.c [hunk 3]
```diff
@@ -728,13 +730,10 @@ void ex_hardcopy(exarg_T *eap)
...
             goto print_fail;
 
-          sprintf((char *)IObuff, _("Printing page %d (%d%%)"),
-              page_count + 1 + side,
-              prtpos.bytes_printed > 1000000
-              ? (int)(prtpos.bytes_printed /
-                      (bytes_to_print / 100))
-              : (int)((prtpos.bytes_printed * 100)
-                      / bytes_to_print));
+          assert(prtpos.bytes_printed * 100 > prtpos.bytes_printed);
```
Comment (fwalch) [inline src/nvim/hardcopy.c]: Won't this always be true?

Comment (gdrooid) [inline src/nvim/hardcopy.c]: Unless `prtpos.bytes_printed` is 0...

Comment (fwalch) [inline src/nvim/hardcopy.c]: Right. But I don't really get the initial code either. Is it trying to avoid overflows?

Comment (elmart) [inline src/nvim/hardcopy.c]: It's not always true. That's a way to check for overflow.

File: src/nvim/hardcopy.c (hunk 2 above)
Comment (elmart) [inline src/nvim/hardcopy.c]: That's true, in principle. I can't carefully review the code now, to see if it has sense doing so. I'll check that tomorrow.

Comment (elmart) [inline src/nvim/hardcopy.c]: Ok, I reviewed every place where INT_MAX was introduced to check conversion to int, and added the corresponding check for INT_MIN when converting from a signed type. Probably, there are cases where, while the variable is signed, it is meant to hold a positive value (mouse_col, mouse_row, etc). In those cases, asserting >= 0 && <= INT_MAX would be more sensible. But, without deeper understanding of the code, and not knowing if value can ever be negative for some reason, I prefer the safe path of asserting >= INT_MIN && <= INT_MAX.

Note also that a lot more of type refactoring cleanup can (and should) still be done in these files, which would eventually eliminate most asserts/casts. But that's not this PR's scope. Here I'm just doing the minimum to be able to enable -Wconversion and remove long_u.

Pushed new version with these changes.

File: src/nvim/hardcopy.c (hunk 3 above)
Comment (elmart) [inline src/nvim/hardcopy.c]: Original code tried to do calculation in two different ways in order to expand the range of values for which calculation could be done without overflow.
Now that prtpos.bytes_printed is size_t, I think range is big enough. So, I simplified calculation to be done just one way, after checking for improbable overflow. I think that's clearer.

File: src/nvim/hardcopy.c (hunk 2 above)
Comment (fwalch) [inline src/nvim/hardcopy.c]: >  Here I'm just doing the minimum to be able to enable -Wconversion and remove long_u.

:+1:

**Comment by justinmk** [comment]
The changes to term.c are pretty noisy, but they don't seem to interfere with #1657.

File: src/nvim/hardcopy.c [hunk 4]
```diff
@@ -358,8 +357,8 @@ static long_u prt_get_term_color(int colorindex)
 static void prt_get_attr(int hl_id, prt_text_attr_T *pattr, int modec)
 {
   int colorindex;
-  long_u fg_color;
-  long_u bg_color;
+  uint32_t fg_color;
+  uint32_t bg_color;
```
Comment (oni-link) [inline src/nvim/hardcopy.c]: Variable `bg_color` is not needed. We could directly set `pattr->bg_color = PRCOLOR_WHITE` at the end of the function.

File: src/nvim/hardcopy.c [hunk 5]
```diff
@@ -3080,16 +3079,18 @@ void mch_print_set_font(int iBold, int iItalic, int iUnderline)
...
   prt_attribute_change = TRUE;
   prt_need_bgcol = TRUE;
 }
 
-void mch_print_set_fg(long_u fgcol)
+void mch_print_set_fg(uint32_t fgcol)
 {
-  if (fgcol != (long_u)prt_fgcol) {
+  assert(fgcol <= INT_MAX);
+  if ((int)fgcol != prt_fgcol) {
```
Comment (oni-link) [inline src/nvim/hardcopy.c]: We could change the type of `prt_fgcol` from `int` to `uint32_t` and drop all casts and range checks.

File: src/nvim/hardcopy.c [hunk 6]
```diff
@@ -3080,16 +3079,18 @@ void mch_print_set_font(int iBold, int iItalic, int iUnderline)
   }
 }
 
-void mch_print_set_bg(long_u bgcol)
+void mch_print_set_bg(uint32_t bgcol)
 {
+  assert(bgcol <= INT_MAX);
```
Comment (oni-link) [inline src/nvim/hardcopy.c]: We could change the type of `prt_bgcol` from `int` to `uint32_t` and drop all casts and range checks.

File: src/nvim/hardcopy.c (hunk 3 above)
Comment (oni-link) [inline src/nvim/hardcopy.c]: `prtpos.bytes_printed` starts with value `0`, so this assert will be always false for the first loop iteration.

Comment (elmart) [inline src/nvim/hardcopy.c]: Ugh, that's true. I'll change > to >=.

File: src/nvim/hardcopy.c (hunk 4 above)
Comment (elmart) [inline src/nvim/hardcopy.c]: :+1: Good catch.

File: src/nvim/hardcopy.c (hunk 6 above)
Comment (elmart) [inline src/nvim/hardcopy.c]: Note that I didn't do all possible type refactoring in these files. There are still a lot of things than can be improved, as I just did the strictly necessary to make -Wconversion pass and remove long_u.
This one is a sensible one, though.

File: src/nvim/hardcopy.c (hunk 5 above)
Comment (elmart) [inline src/nvim/hardcopy.c]: Same as above.

**Comment by elmart** [comment]
@oni-link I will address all your comments in a subsequent PR. Thanks!

**Pull request merged**